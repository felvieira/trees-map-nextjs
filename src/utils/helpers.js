import { areaColors } from "styles/colors";

export const convertHexToRGBA = (hexCode, opacity) => {
    let hex = hexCode.replace('#', '');
    
    if (hex.length === 3) {
        hex = `${hex[0]}${hex[0]}${hex[1]}${hex[1]}${hex[2]}${hex[2]}`;
    }    
    
    const r = parseInt(hex.substring(0, 2), 16);
    const g = parseInt(hex.substring(2, 4), 16);
    const b = parseInt(hex.substring(4, 6), 16);

    return `rgba(${r},${g},${b},${opacity / 100})`;
};

export const getCenterOfPolygon = function (arr) {
  var twoTimesSignedArea = 0;
  var cxTimes6SignedArea = 0;
  var cyTimes6SignedArea = 0;

  var length = arr.length;

  var x = function (i) {
    return arr[i % length][0];
  };
  var y = function (i) {
    return arr[i % length][1];
  };

  for (var i = 0; i < arr.length; i++) {
    var twoSA = x(i) * y(i + 1) - x(i + 1) * y(i);
    twoTimesSignedArea += twoSA;
    cxTimes6SignedArea += (x(i) + x(i + 1)) * twoSA;
    cyTimes6SignedArea += (y(i) + y(i + 1)) * twoSA;
  }
  var sixSignedArea = 3 * twoTimesSignedArea;
  return [
    cxTimes6SignedArea / sixSignedArea,
    cyTimes6SignedArea / sixSignedArea,
  ];
};

export const getPopulation = (arr, id, year) =>
  arr[id - 1].population.filter((pop) => pop.ano === year)[0].populacao;

export const getIconMarkerByPopulation = (pop) => {
  pop = parseInt(pop);
  switch (true) {
    case pop <= 8000:
      return {
        src: "./img/pop7k-8k.png",
        size: "70px",
      };
    case pop <= 11000:
      return {
        src: "./img/pop8k-11k.png",
        size: "60px",
      };
    case pop <= 13000:
      return {
        src: "./img/pop11k-13k.png",
        size: "70px",
      };
    case pop <= 18000:
      return {
        src: "./img/pop13k-18k.png",
        size: "90px",
      };
    case pop <= 21000:
      return {
        src: "./img/pop18k-21k.png",
        size: "140px",
      };
    default:
      return {
        src: "./img/pop18k-21k.png",
        size: "130px",
      };
  }
};

export const styledMapFn = (i) => {
  return {
    id: `style-map-${i}`,
    type: "fill",
    paint: {
      "fill-color": areaColors[i],
      "fill-opacity": 1,
    },
  };
};
