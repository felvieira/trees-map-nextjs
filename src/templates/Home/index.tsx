import { useState } from 'react'
import ReactMapGL from 'react-map-gl'
import { getCenterOfPolygon } from 'utils/helpers'
import MenuYears from 'components/MenuYear'
import MenuChart from 'components/MenuChart'
import WrapperMaps from 'components/WrapperMaps'

export type MainProps = {
	description?: string
}

const Home = ({ datas, years }) => {
	const [viewport, setViewport] = useState({
		latitude: 0.751348,
		longitude: -68.578,
		zoom: 11
	})

	return (
		<>
			<MenuYears years={years} />
			<ReactMapGL
				{...viewport}
				width="100vw"
				height="100vh"
				mapStyle="mapbox://styles/flpchapola/cksw8w3ki161d1anqdenv0a0j"
				mapboxApiAccessToken={process.env.mapBox}
				onViewportChange={setViewport}
			>
				{datas &&
					datas.map((item, i) => {
						const coords = item.geometry.coordinates[0]
						const latitude = getCenterOfPolygon(coords)[1]
						const longitude = getCenterOfPolygon(coords)[0]

						return (
							<div key={`wrapper-maps-${i}`}>
								<WrapperMaps
									data={datas}
									item={item}
									index={i}
									latitude={latitude}
									longitude={longitude}
								/>
								<MenuChart data={datas} years={years} />
							</div>
						)
					})}
			</ReactMapGL>
		</>
	)
}

export default Home
