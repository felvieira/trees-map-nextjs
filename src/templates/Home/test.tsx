import { renderWithTheme } from 'utils/tests/helpers'

import Home from '.'

describe('<Home />', () => {
	it('should render the MAIN page', () => {
		const { container } = renderWithTheme(<Home />)

		expect(container.firstChild).toMatchSnapshot()
		expect(container.firstChild).toBeTruthy()
	})
})
