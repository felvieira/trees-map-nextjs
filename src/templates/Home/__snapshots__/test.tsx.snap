// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render the MAIN page 1`] = `
.c0 {
  z-index: 1;
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  width: 100%;
  padding: 1.6rem;
  background: #7FB685;
  border-bottom: 6px solid #426A5A;
}

.c1 {
  background: #426A5A;
  padding: 0.8rem 1.6rem;
  border-radius: 0.5rem;
  font-weight: 700;
  color: #7FB685;
  cursor: pointer;
}

.c2 {
  background: none;
  padding: 0.8rem 1.6rem;
  border-radius: 0.5rem;
  font-weight: 700;
  color: #426A5A;
  cursor: pointer;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    2000
  </div>
  <div
    class="c2"
  >
    2002
  </div>
  <div
    class="c2"
  >
    2004
  </div>
  <div
    class="c2"
  >
    2006
  </div>
</div>
`;
